core/src/syscall/host.rs

```diff
@@ -202,9 +202,17 @@ impl Syscall for SyscallHostSpawn {
     let vec_words = ctx.slice(address, len_words);
     let blob = vec_u32_to_bytes(vec_words, len as usize);
 
-    // get value from host
     let host = ctx.rt.host.as_deref_mut().expect("Missing host interface");
-    host.spawn(blob);
+    let address = host.spawn(blob);
+
+    let out_addr = ctx
```

Comment:
Checking memory safety here. Are we sure that a user program cannot sneak in a different memory location here before this write happens? Worth adding a comment here explaining why, it's not immediately obvious.

Reply:
What do you mean by `user program`? Do you mean the guest (the risc-v program)? This one is halt for the entire time of the syscall so it cannot do anything. The `out_addr` and the memory it points to only exist in the context of this `Runtime` instance (this is basically a map in the `Runtime` struct). 

https://github.com/athenavm/athena/blob/8f4dd9ef8839c297b6d9a4238c1c645f829c1e44/core/src/runtime/state.rs#L22

It's not possible to overwrite the host memory this way. Check out the `mw` routine: https://github.com/athenavm/athena/blob/ae54d14c7c8b489240765cafb393d5266665cdc6/core/src/runtime/mod.rs#L222-L237

Reply:
No, I mean a guest program overwriting the memory of another guest program (sorry if I wasn't clear). E.g., imagine a smart contract calling into another smart contract: the first execution frame is paused while the second runs. Depending how this is implemented, they could share memory. (The way it's built today is quite naive/inefficient, an entirely new VM is spun up for the call, so I'm pretty sure they don't share memory, but this might change in the future).

Alternatively: a guest program overwriting its own code. For safety we probably want to maintain the constraint that a program cannot overwrite its own instructions (see https://en.wikipedia.org/wiki/Harvard_architecture).

Reply:
> No, I mean a guest program overwriting the memory of another guest program (sorry if I wasn't clear). E.g., imagine a smart contract calling into another smart contract: the first execution frame is paused while the second runs. Depending how this is implemented, they could share memory. (The way it's built today is quite naive/inefficient, an entirely new VM is spun up for the call, so I'm pretty sure they don't share memory, but this might change in the future).

OK, I understand now. You are right, it would be a problem if they shared the memory. But the way the VM `Runtime` is currently implemented, running 2 programs in the same `Runtime` isn't safe - they could easily manipulate the memory of one another by executing store instructions and writing the memory space used by another. The way to separate memory spaces of 2 programs "physically" is to run separate Runtimes.

> Alternatively: a guest program overwriting its own code. For safety we probably want to maintain the constraint that a program cannot overwrite its own instructions (see [en.wikipedia.org/wiki/Harvard_architecture](https://en.wikipedia.org/wiki/Harvard_architecture)).

The program instructions and memory are separated and it's impossible to overwrite the program: https://github.com/athenavm/athena/blob/096857aa5b37e38f466f5c01834da161ce69c06b/core/src/runtime/mod.rs#L38-L46

The program memory is held in `state.memory` separately from the instructions and writing it cannot modify them.